### PR TITLE
Add transaction circuit

### LINKED ISSUE #169: Merge transaction circuits into one
#### Summary
Currently the circuits that need to be proven to create a transaction in zk are split into multiple parts: the execute circuits, send to contract and withdraw from contract. To simplify transactions, possibly reduce the gate counts and speed up proof creation, we want to merge the transfer circuit into a singular circuit based on the diagram in the protocol specifications.

#### Possible Solution
With this effort we also want to add the circuit directly in phoenix-core, as it is part of the phoenix protocol.

### PR DESCRIPTION
**Pull request by xevisalle**

### CODE REVIEW

**Review by moCello — APPROVED** [review]
LGTM

File: circuits/README.md [hunk 1]
```diff
@@ -8,4 +8,4 @@ This library contains the implementation of the Phoenix-circuits, to prove, in z
 2. Ownership: the sender holds the note secret key for every note that is about to be spent.
 3. Nullification: the nullifier is calculated correctly.
 4. Minting: the value commitment for the newly minted notes are computed correctly.
-5. Balance integrity: the sum of the value of all spent notes is equal to the value of the sum of all minted notes + the gas fee + a crossover, where a crossover refers to funds being transfered to a contract.
+5. Balance integrity: the sum of the values of all spent notes is equal to the sum of the values of all minted notes + the gas fee + a crossover, where a crossover refers to funds being transfered to a contract.
```
Comment (ureeves) [inline circuits/README.md:11]: I understand what you're trying to emphasize here, but the previous sentence feels more correct on the english front.

File: circuits/src/transaction.rs [hunk 2]
```diff
@@ -0,0 +1,419 @@
...
+use alloc::vec::Vec;
+
+use phoenix_core::{Error as PhoenixError, Note, Ownable, SecretKey, ViewKey};
+
+const TX_OUTPUT_NOTES: usize = 2;
+
+/// Struct representing a note willing to be spent, in a way
+/// suitable for being introduced in the transfer circuit
+#[derive(Debug, Clone)]
+pub struct TxInputNote<const H: usize, const A: usize> {
```
Comment (ureeves) [inline circuits/src/transaction.rs:28]: I feel like this should be called `TxInput`

File: circuits/src/transaction.rs [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+    pub(crate) note: Note,
+    pub(crate) note_pk_p: JubJubAffine,
+    pub(crate) value: u64,
+    pub(crate) blinding_factor: JubJubScalar,
+    pub(crate) nullifier: BlsScalar,
+    pub(crate) signature: SignatureDouble,
+}
+
+#[derive(Debug, Clone)]
+struct WitnessTxInputNote {
```
Comment (ureeves) [inline circuits/src/transaction.rs:39]: And this should be called `TxInputWitness`

File: circuits/src/transaction.rs [hunk 4]
```diff
@@ -0,0 +1,419 @@
...
+            signature_r,
+            signature_r_p,
+        }
+    }
+}
+
+/// Struct representing a note willing to be created, in a way
+/// suitable for being introduced in the transfer circuit
+#[derive(Debug, Clone)]
+pub struct TxOutputNote {
```
Comment (ureeves) [inline circuits/src/transaction.rs:127]: As well as this one being called `TxOutput`

File: circuits/src/transaction.rs [hunk 5]
```diff
@@ -0,0 +1,419 @@
...
+    pub(crate) value: u64,
+    pub(crate) value_commitment: JubJubAffine,
+    pub(crate) blinding_factor: JubJubScalar,
+}
+
+#[derive(Debug, Clone)]
+struct WitnessTxOutputNote {
```
Comment (ureeves) [inline circuits/src/transaction.rs:134]: And this being called `TxOutputWitness`

**Review by ureeves — APPROVED** [review]
All I have is naming considerations, which we've discussed, and one nit on the change in the README. Otherwise LGTM. Please do squash your commits though

**Pull request merged**